### PR TITLE
Save/restore TransactionView and recentRequestsView tables column sizes

### PR DESCRIPTION
**Pull request by hebasto**
Save/restore TransactionView and recentRequestsView tables column sizes.
Sorting order is not saved/restored intentionally.

Based on #204 (the first commit).

### CODE REVIEW

**Comment by jonatack** [comment]
Concept ACK if the idea is to save the column sizes if adjusted by the user and restore them on next use (is that the goal of this pull?), ATM for me on Debian the Transactions window column widths are indeed not persisted after shutdown. (Why not save the sorting order too?)

**Comment by hebasto** [comment]
> ... if the idea is to save the column sizes if adjusted by the user and restore them on next use (is that the goal of this pull?)...

Correct.

> Why not save the sorting order too?

After a while users could forget/unnotice another sort order, and that could cause their panic (when they are not looking at the recent made transaction).

**Comment by jonatack** [comment]
Tested, it seems to work well. Sort is not preserved but column widths are, as described.

**Comment by jonasschnelli** [comment]
Concept ACK (I guess this is also resettable by `-resetguisettings`?)

**Comment by hebasto** [comment]
@jonasschnelli 
> I guess this is also resettable by `-resetguisettings`?

Correct. `-resetguisettings` affects all of the settings those use a `QSettings` class instance.

**Review by jarolrod — COMMENTED** [review]
ACK 964885d04801c6ab77ce4705cff01c9d83bc3ed8, tested on macOS 11.1 Qt 5.15.2

PR works well and does what it says it will do. Confirmed that the column width's of transactions and recent requests are saved and then restored upon node restart. Code looks good and it removes some complexity (yay, less lines of code).

Below are screenshots comparing the behavior between `master` and this `pr`

**Master**: Start point
MASTER-Start

**Master**: Resize
MASTER-resize

**Master**: Restart, column width does not persist
MASTER-does-not-persist

**PR**: Start
PR-start

**PR:** Resize
PR-resize

**PR**: Restart, column width persists
PR-persists

File: src/qt/transactionview.cpp [hunk 1]
```diff
@@ -126,27 +127,40 @@ TransactionView::TransactionView(const PlatformStyle *platformStyle, QWidget *pa
...
+    transactionView->setContextMenuPolicy(Qt::CustomContextMenu);
+    transactionView->installEventFilter(this);
+    transactionView->setAlternatingRowColors(true);
+    transactionView->setSelectionBehavior(QAbstractItemView::SelectRows);
+    transactionView->setSelectionMode(QAbstractItemView::ExtendedSelection);
+    transactionView->setSortingEnabled(true);
+    transactionView->verticalHeader()->hide();
+
+    QSettings settings;
+    if (!transactionView->horizontalHeader()->restoreState(settings.value("TransactionViewHeaderState").toByteArray())) {
```
Comment (Talkless) [inline src/qt/transactionview.cpp:154]: tiny nit:
`TransactionViewHeaderState` string literal is used twice in same .cpp file. Maybe we can then consider this as "magic string" (as magic constant) and avoid duplicating, using some local .cpp-file constant?

**Review by Talkless — APPROVED** [review]
tACK 964885d04801c6ab77ce4705cff01c9d83bc3ed8, tested on Debian Sid, saving/restoring and resetting (with `-resetguisettings`) works as expected.

File: src/qt/transactionview.cpp (hunk 1 above)
Comment (hebasto) [reply src/qt/transactionview.cpp:154]: This is the current way to work with setting names in the codebase.
Your suggestion is nice to apply to _all_ of the setting names in a follow up pr.

Comment (Talkless) [reply src/qt/transactionview.cpp:154]: > Your suggestion is nice to apply to all of the setting names in a follow up pr.

True. Maybe even `GUISettings` class to wrap calls to `QSettings` with methods like `getTransactionViewHeaderState()`.

**Comment by Geremia** [comment]
This fixed the column widths issue #21306; but now, after opening bitcoin-qt, the transactions aren't chronologically sorted until I click the date column header twice.

**Comment by hebasto** [comment]
@Geremia 
> but now, after opening bitcoin-qt, the transactions aren't chronologically sorted until I click the date column header twice.

Thank you for reporting. Fixed in #229.

> Also, this doesn't fix "Requested payments history" and "Peers" column widths issues.

Are you sure about the "Requested payments history" tab?

The "Peers" tab is WIP :)

**Comment by Geremia** [comment]
@hebasto 
>Are you sure about the "Requested payments history" tab?

Yes, widths aren't remembered.

**Comment by hebasto** [comment]
While working on #256, two flaws were noted in this PR.

1. In multi-wallet environment settings of the transaction table of the last closed wallet are saved only.
2. `QTableView::setColumnWidth` does not work without model set (but it is not documented by Qt).

:disappointed:

**Pull request merged**